### PR TITLE
Add a platform specific configuration checker

### LINKED ISSUE #2846: MinGW build fails
Hello,

I'm facing the following issue building 1.1.0e using MSYS from [MinGW.org](http://mingw.org), and F[ree Community ActivePerl](https://www.activestate.com/activeperl).
No issue with 1.0.2k under the same build environment.

Thank you 👍 

Ben

```
$ ./config no-asm no-shared
Operating system: i686-whatever-mingw
Configuring for mingw
Configuring OpenSSL version 1.1.0e (0x1010005fL)
    no-asan         [default]  OPENSSL_NO_ASAN
    no-asm          [option]   OPENSSL_NO_ASM
    no-crypto-mdebug [default]  OPENSSL_NO_CRYPTO_MDEBUG
    no-crypto-mdebug-backtrace [default]  OPENSSL_NO_CRYPTO_MDEBUG_BACKTRACE
    no-dynamic-engine [forced]
    no-ec_nistp_64_gcc_128 [default]  OPENSSL_NO_EC_NISTP_64_GCC_128
    no-egd          [default]  OPENSSL_NO_EGD
    no-fuzz-afl     [default]  OPENSSL_NO_FUZZ_AFL
    no-fuzz-libfuzzer [default]  OPENSSL_NO_FUZZ_LIBFUZZER
    no-heartbeats   [default]  OPENSSL_NO_HEARTBEATS
    no-md2          [default]  OPENSSL_NO_MD2 (skip dir)
    no-msan         [default]  OPENSSL_NO_MSAN
    no-rc5          [default]  OPENSSL_NO_RC5 (skip dir)
    no-sctp         [default]  OPENSSL_NO_SCTP
    no-shared       [option]
    no-ssl-trace    [default]  OPENSSL_NO_SSL_TRACE
    no-ssl3         [default]  OPENSSL_NO_SSL3
    no-ssl3-method  [default]  OPENSSL_NO_SSL3_METHOD
    no-ubsan        [default]  OPENSSL_NO_UBSAN
    no-unit-test    [default]  OPENSSL_NO_UNIT_TEST
    no-weak-ssl-ciphers [default]  OPENSSL_NO_WEAK_SSL_CIPHERS
    no-zlib         [default]
    no-zlib-dynamic [default]
Configuring for mingw
CC            =gcc
CFLAG         =-DL_ENDIAN -DWIN32_LEAN_AND_MEAN -DUNICODE -D_UNICODE -m32 -Wall -O3 -fomit-frame-pointer -D_MT
SHARED_CFLAG  =-D_WINDLL
DEFINES       =DSO_WIN32 NDEBUG OPENSSL_THREADS OPENSSL_NO_DYNAMIC_ENGINE OPENSSL_PIC
LFLAG         =
PLIB_LFLAG    =
EX_LIBS       =-lws2_32 -lgdi32 -lcrypt32
APPS_OBJ      =win32_init.o
CPUID_OBJ     =mem_clr.o
UPLINK_OBJ    =
BN_ASM        =bn_asm.o
EC_ASM        =
DES_ENC       =des_enc.o fcrypt_b.o
AES_ENC       =aes_core.o aes_cbc.o
BF_ENC        =bf_enc.o
CAST_ENC      =c_enc.o
RC4_ENC       =rc4_enc.o rc4_skey.o
RC5_ENC       =rc5_enc.o
MD5_OBJ_ASM   =
SHA1_OBJ_ASM  =
RMD160_OBJ_ASM=
CMLL_ENC      =camellia.o cmll_misc.o cmll_cbc.o
MODES_OBJ     =
PADLOCK_OBJ   =
CHACHA_ENC    =chacha_enc.o
POLY1305_OBJ  =
BLAKE2_OBJ    =
PROCESSOR     =
RANLIB        =ranlib
ARFLAGS       =
PERL          =c:\Perl\bin\perl.exe

THIRTY_TWO_BIT mode
BN_LLONG mode

Configured for mingw.

$ make
make[1]: Entering directory `/e/src/packages/openssl-1.1.0e'
c:\Perl\bin\perl.exe "-I." -Mconfigdata "util\dofile.pl" \
            "-oMakefile" crypto\include\internal\bn_conf.h.in > crypto\include\internal\bn_conf.h
/bin/sh: c:Perlbinperl.exe: command not found
make[1]: *** [crypto\include\internal\bn_conf.h] Error 127
make[1]: Leaving directory `/e/src/packages/openssl-1.1.0e'
make: *** [curl] Error 2
```

Forcing `PERL = perl.exe` in Makefile :

```
$ make
perl.exe "-I." -Mconfigdata "util\dofile.pl" \
            "-oMakefile" crypto\include\internal\bn_conf.h.in > crypto\include\internal\bn_conf.h
Couldn't open file cryptoincludeinternalbn_conf.h.in: No such file or directory
make: *** [crypto\include\internal\bn_conf.h] Error 2
```

Generated Makefile contains paths with `/` separator, others with `\` separator...
Certainly `\` should not be used ?

### PR DESCRIPTION
**Pull request by levitte**
For each platform, we may need to perform some basic checks to see
that available tools perform as we expect them.

For the moment, the added checkers test that Perl gives the expected
path format.  This should help MingW users to see if they run an
appropriate Perl implementation, for example.

##### Checklist

- [ ] documentation is added or updated

-----

Fixes #2846

### CODE REVIEW

**Comment by levitte** [comment]
Documentation needs added, that's why it's still WIP.  However, I'd appreciate some early comments.

**Comment by richsalz** [comment]
What docs, other than a README?  But this looks good so far.

File: Configurations/unix-checker.pm [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+# Check that the perl implementation file modules generate paths that
+# we expect for the platform
+use File::Spec::Functions qw(:DEFAULT rel2abs);
+
+if (rel2abs('.') !~ m|/|) {
+    die <<EOF;
+
+******************************************************************************
+This perl version doesn't produce Unix like paths (with forward slash
+directory separators).  Please use an implementation that does.
```
Comment (mattcaswell) [inline Configurations/unix-checker.pm]: Perhaps something about using "matched" perl versions might be a useful hint?

**Comment by levitte** [comment]
Documentation added.  (yes, @richsalz, the README ;-) )

**Comment by levitte** [comment]
@mattcaswell, I just pushed an update...  is the text better now?

File: Configure [hunk 2]
```diff
@@ -993,6 +993,25 @@ $target{build_scheme} = [ $target{build_scheme} ]
 my ($builder, $builder_platform, @builder_opts) =
     @{$target{build_scheme}};
 
+foreach my $checker (($builder_platform."-".$target{build_file}."-checker.pm",
+                      $builder_platform."-checker.pm")) {
+    my $checker_path = catfile($srcdir, "Configurations", $checker);
+    if (-f $checker_path) {
+        my $fn = $ENV{CONFIGURE_CHECKER_WARN}
```
Comment (mattcaswell) [inline Configure]: Why is this necessary?

**Comment by dot-asm** [comment]
Is it possible to place it directly to `*-makefile.tmpl`? After all, the constraint  has *everything* to do with the target makefile...

**Comment by dot-asm** [comment]
> This should help MingW users to see if they run an appropriate Perl implementation, for example.

It only helps to detect obviously inappropriate combination, not all incompatible ones. Well, I'm not actually suggesting to attempt the latter. As it's impossible to formalize...

File: Configure (hunk 2 above)
Comment (levitte) [inline Configure]: It's for the "I fscking know what I'm doing" mode.  For example, it happens that I do `./Configure VC-WIN32` on my Linux box just to check certain aspects of `configdata.pm` without having to start a VirtualBox instance runnign Windows...

**Comment by levitte** [comment]
> Is it possible to place it directly to `*-makefile.tmpl`? After all, the constraint  has *everything* to do with the target makefile...

I was thinking about it, but that seemed overly complex, I wasn't sure of all the consequences.

**Comment by dot-asm** [comment]
> > Is it possible to place it directly to *-makefile.tmpl?
>
> I was thinking about it, but that seemed overly complex

But there are blocks in .tmpl-ates that are evaluated. And if I simply drop die there, it dies... And it's last line. Well, an empty Makefile is created, and one can argue that it shouldn't, i.e. test should be performed earlier... Oh well...

File: Configure [hunk 3]
```diff
@@ -993,6 +993,25 @@ $target{build_scheme} = [ $target{build_scheme} ]
...
+            ? sub { warn $@; } : sub { die $@; };
+        if (! do $checker_path) {
+            if ($@) {
+                $fn->($@);
+            } elsif ($!) {
+                $fn->($!);
```
Comment (dot-asm) [inline Configure]: Is `errno` actually meaningful here?

**Comment by levitte** [comment]
> Well, an empty Makefile is created, and one can argue that it shouldn't, i.e. test should be performed earlier... Oh well...

Exactly.

File: Configure (hunk 3 above)
Comment (levitte) [inline Configure]: Not right now, but you know, "belt and braces" 'n all that...

**Comment by levitte** [comment]
Another thing, apropos implementing this in the build file templates, is that the check is made relatively late in the configuration process.  The idea with this check is to have it fairly early...  to "nip it in the bud", so to say.

File: Configure (hunk 3 above)
Comment (dot-asm) [inline Configure]: And concern is that `errno` can originate from some unrelated operation and printing it here would be misleading.

Comment (dot-asm) [inline Configure]: Oh maybe you mean that stat(".") implicitly called in rel2abs would set errno to something funny? Like "access denied" or "no such file" if you delete current working directory? ...

Comment (levitte) [inline Configure]: No, I've followed what is documented about `do`.  From http://perldoc.perl.org/functions/do.html

```
If do can read the file but cannot compile it, it returns undef and sets an error message in $@ . If do cannot read the file, it returns undef and sets $! to the error. Always check $@ first, as compilation could fail in a way that also sets $! . If the file is successfully compiled, do returns the value of the last expression evaluated.
```

Comment (dot-asm) [inline Configure]: Ah! In other words `errno` is in fact meaningful.

**Review by dot-asm — APPROVED** [review]
+1

**Comment by dot-asm** [comment]
On hardly related note, Strawberry Perl is also an option for Windows builds.

**Comment by levitte** [comment]
Merged.

master: d192a3aaeb76fc89f8285b4dc938c2bc0c37d0d4 and 44eb65ce20d673d0332802275d54f6811f448076
1.1.0: a40b5be27e28eaf5a8be7193c8b6460bc6f851b4 and 08beac4f3d37247e2ed3bf49dd7d760f4c3e0389

**Comment by levitte** [comment]
> Strawberry Perl is also an option for Windows builds.

Have we tested it?  Feel free to add something about it in `NOTES.PERL`
